### PR TITLE
Update Build System

### LINKED ISSUE #1320: Consider the local development process with docker containers in the new versioning system
With the new versioning system, the version will change when a commit is made. If you make a commit and then launch a scenario, it will error because there isn't a docker container for it.  It seems onerous to me to rebuild the docker image (even if it is only the last couple of layers / armory install) each time I locally add a commit. On the other hand, I would guess that 99% of the time if I simply used the previous docker image and ran from the armory directory, it would be fine. I would certainly want the CI test to build a new container, but not something I want to do every time I commit something.

Thoughts?

### PR DESCRIPTION
**Pull request by christopherwoodall**
Update Build System to support [PEP-0518](https://peps.python.org/pep-0518/).

Changes include:
  - Consolidating setup and configuration into `pyproject.toml`.
  - Introducing `hatch` as the package builder.
  - Add build hooks to update `__about__.py` with the appropriate VCS tag.
  - Updating `version.py` to facilitate the above change.

The new versioning scheme is almost what we want modulo some pip demands on version strings.

    $ armory --version
    0.15.4+buildc22a63f

Where `c22a63f` is the head commit. As always if you are at a tagged head there is no commit tag like `0.16.0`. I've
explained the mechanism in [the armory faq](docs/faqs.md).

With the advent of the pyproject, **all** dependencies and versions are in there. So `pip install armory-testbed` gets you a frameworkless installation, while `pip install armory-testbed[pytorch]` gets you that flavor. There are other optional dependencies like `jupyterlab`, `developer` and even `all`.  These can be freely combined and are documented in [Getting Started](docs/getting_started.md#the-armory-flavors). 

Fixes #1320, #1342, #1569, #1581, #1593

### CODE REVIEW

File: armory/__about__.py
Comment (davidslater) [inline armory/__about__.py:1]: Hard coding this seems to go completely against the reason for SCM versioning.

Comment (davidslater) [reply armory/__about__.py:1]: How will this update when a new commit is added?

File: armory/utils/version.py
Comment (davidslater) [inline armory/utils/version.py:38]: The `:=` operator was introduced in python 3.8, iirc. Adding it here would break our code for python 3.7.

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:43]: https://armory.readthedocs.io/en/latest/

**Comment by davidslater** [comment]
What does `hatch` do?

File: armory/utils/version.py
Comment (davidslater) [reply armory/utils/version.py:38]: Yep, fails for me in python 3.7:
```
>>> import armory
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david.slater/git/davidslater/armory/armory/__init__.py", line 6, in <module>
    from armory.utils import version, typedef
  File "/Users/david.slater/git/davidslater/armory/armory/utils/version.py", line 26
    if (tag_match := git_tag_regex.match(version_str)) is not None:
                  ^
SyntaxError: invalid syntax
```

**Comment by davidslater** [comment]
The armory version is wrong, when I grab it, unless I misunderstand how you're putting things together.
```
>>> import armory
>>> armory.__version__
'0.15.5'
```

File: armory/utils/version.py
Comment (christopherwoodall) [reply armory/utils/version.py:38]: The `__about__.py` file is modified dynamically using a [build hook](https://github.com/christopherwoodall/armory/blob/976905143a50f9f128190b4ef5042e4a9bce74de/pyproject.toml#L141), so that anytime the package is installed- locally- the `VCS_VERSION` will update. That said, once the package is built- e.g. sdist/wheel- that version is then static.

Since the `.git` directory is not included in the distribution and is specifically ignored in the [`.dockerignore`](https://github.com/christopherwoodall/armory/blob/976905143a50f9f128190b4ef5042e4a9bce74de/.dockerignore#L2), once published, there is no other way to retrieve the release tag unless it is written somewhere.  

Thank you for catching the walrus ops, I bumped the Python version to 3.8 not realizing there was still code with specific support for 3.7.

Comment (davidslater) [reply armory/utils/version.py:38]: How does the versioning work with editable pip installs?

**Comment by christopherwoodall** [comment]
> \>\>\> armory.__version__
> \>\>\> '0.15.5'

Just noticed that as well. [It](https://hatch.pypa.io/latest/plugins/version-source/reference/) is incrementing the assigned tag- `git describe --always --abbrev=0`. I will look into cloning the plugin locally and modifying it.

File: armory/__about__.py
Comment (christopherwoodall) [reply armory/__about__.py:1]: I [replied here](https://github.com/twosixlabs/armory/pull/1617#discussion_r976677381) as the comments are in a similar vein.

**Comment by ofek** [comment]
Hatch maintainer here! Let me know if I can help

File: docker/Dockerfile-pytorch
Comment (davidslater) [inline docker/Dockerfile-pytorch:18]: Where does this tar file come from? Is it set up somewhere else?

File: .flake8
Comment (davidslater) [inline .flake8:6]: What does this do?

File: .flake8
Comment (davidslater) [inline .flake8:22]: What additional items did this add?

File: tools/pre-commit.sh
Comment (davidslater) [inline tools/pre-commit.sh:26]: Where did yamllint go?

File: tools/pre-commit.sh
Comment (davidslater) [inline tools/pre-commit.sh:1]: Can you describe the changes made in this file? I'm having a hard time following all of them.

File: armory/utils/version.py
Comment (christopherwoodall) [reply armory/utils/version.py]: > How does the versioning work with editable pip installs?

When building containers, the version is pulled when [the package is being built](https://github.com/christopherwoodall/armory/blob/develop/docker/build.py#L80); tag is pulled from the local `git` repository then added to the "metadata", and finally included in the sdist.

On a local machine, the tags are pulled directly from the `.git` metadata on every `pip install -e `. 


One minor change to the pip install is that instead of specifying a requirements file- e.g. `pip install -e host-requirements.txt`, the developer would instead run `pip install -e '.[developer,host]'` depending on what tools they need.

File: pyproject.toml
Comment (christopherwoodall) [reply pyproject.toml]: Updated [`pyproject.toml`](https://github.com/christopherwoodall/armory/blob/develop/pyproject.toml#L45).

File: docker/Dockerfile-pytorch
Comment (christopherwoodall) [reply docker/Dockerfile-pytorch]: The tar file(`sdist`) is the output from the build process. This happens in [`build.py`](https://github.com/christopherwoodall/armory/blob/develop/docker/build.py#L80) using `hatch`, but could easily switch to `python -m build`.

In saying that, I've modified the above approach. The [new approach](https://github.com/christopherwoodall/armory/actions/runs/3138111637) adds an exemption for the `.git` directory in [`.dockerignore`](https://github.com/christopherwoodall/armory/blob/build-system/.dockerignore), then uses a single [`COPY` command](https://github.com/christopherwoodall/armory/blob/build-system/docker/Dockerfile-pytorch#L10) to copy the necessary metadata for a local(in container) install.

File: .flake8
Comment (christopherwoodall) [reply .flake8:6]: Those are arguments for the linter. (WIP)

File: .flake8
Comment (christopherwoodall) [reply .flake8:20]: Reverted and updated.

File: tools/pre-commit.sh
Comment (christopherwoodall) [reply tools/pre-commit.sh:1]: Reverted. I will throw add this in the CI/Actions branch.

File: tools/pre-commit.sh
Comment (christopherwoodall) [reply tools/pre-commit.sh:26]: Reverted. Will add to the Actions/CI branch.

**Review by christopherwoodall — COMMENTED** [review]
The requested changes have been implemented.

File: docker/Dockerfile-pytorch
Comment (davidslater) [inline docker/Dockerfile-pytorch:16]: May revisit the `[host]` addition.

File: docker/build.py
Comment (davidslater) [inline docker/build.py:87]:
```
(armory)david.slater@noether:~/armory$ git describe
v0.15.4-83-g2906e99
(armory)david.slater@noether:~/armory$ armory --version
0.15.5.g2906e99
```

For tagging docker container, see armory version above.

File: docker/build.py
Comment (davidslater) [inline docker/build.py:77]: Old code?

File: docker/build.py
Comment (davidslater) [inline docker/build.py:15]: Add a note: # podman is not officially supported by armory, but his enables use as a drop-in replacement for building.
(OR: container_platform is an argument that defaults to "docker", but if you wanted a drop-in replacement, do --container_platform=your_replacement.)

File: armory/utils/version.py
Comment (davidslater) [inline armory/utils/version.py:79]: setuptools?

File: armory/__init__.py
Comment (davidslater) [inline armory/__init__.py:14]: Only get version if they ask for it.

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:175]: Matt will review hatch.

File: docker/build.py
Comment (christopherwoodall) [reply docker/build.py:19]: Added comment

File: docker/build.py
Comment (christopherwoodall) [reply docker/build.py]: Fixed using `scm` from `setuptools`. Modified `pyproject.toml`, `version.py`, and `build.py` to all use `setuptools_scm` along with the same formatting arguments.

Developers are likely to experience version mismatches if a file is edited while/during the building of an image due to the versioning requirements, highly dynamic/volatile nature of the file system during development, and the interdependence of `armory` on `docker`(due, in part, to the [available configuration options](https://github.com/pypa/setuptools_scm/blob/c4d37004ab0a16c2cacdc58ef06b36956db02a9f/setup.cfg) within `scm` as well as `docker` tag syntax. ).

File: docker/build.py
Comment (christopherwoodall) [reply docker/build.py]: Removed lingering code.

File: armory/utils/version.py
Comment (christopherwoodall) [reply armory/utils/version.py]: Implemented setuptools, however, this will introduce a new set of problems for people editing files and testing at the same time. A workaround would be the manual retagging of locally built images, and a possible enhancement could include adding an entrypoint to `armory.scenarios.main` to enable direct interaction with `armory` from within the image.

File: armory/__init__.py
Comment (christopherwoodall) [reply armory/__init__.py]: Implemented `__getattr__`.

File: tests/conftest.py
Comment (christopherwoodall) [reply tests/conftest.py]: Reverted.

File: docker/Dockerfile-pytorch
Comment (christopherwoodall) [reply docker/Dockerfile-pytorch:16]: I agree. I came across this [blog](https://ealizadeh.com/blog/guide-to-python-env-pkg-dependency-using-conda-poetry) that explains the rationale better than I can.

**Comment by mwartell** [comment]
`README.md` was moved from / to `.github`.  I'm pretty sure that breaks our github front page. Was there a reason for this?

**Comment by mwartell** [comment]
> `README.md` was moved from / to `.github`. I'm pretty sure that breaks our github front page. Was there a reason for this?

As discussed, this is actually a valid place to put a readme (surprise). nI asked you to move it back because it violated expectations of 
```
git clone armory
cat armory/README.md
```

**Comment by mwartell** [comment]
Here's my explanation of what I think the pyproject dependencies should look. I'd like a user to be able to say:

    pip install armory-testbed    # loads things that are required to run the launcher and spawn the container (maybe ART?)
    pip install armory-testbed[tensorflow]  # base + tensorflow libraries
    pip install armory-testbed[pytorch]    # base + pytorch + tfds
    pip install armory-testbed[pytorch-deepspeech] # base + pytorch + tfds + deepspeech

Which means pyproject would look like 
```
[project.optional-dependencies]
tensorflow = [
    …
]
pytorch = [...]
pytorch-deepspeech = [...]
```
Toml does not allow recursive optional dependencies, but we could have a command like
```
pip install armory-testbed[pytorch,deepspeech] 
``` 
which should help

**Comment by mwartell** [comment]
The proper package name for pyproject.toml is  "armory-testbed" because of PyPI uniqueness requirements.

**Review by mwartell — APPROVED** [review]
with a few days active testing and debugging this, LGTM

**Comment by christopherwoodall** [comment]
If merged, this PR will close the following issues:

- [Consider the local development process with docker containers in the new versioning system](https://github.com/twosixlabs/armory/issues/1320)
- [validate that python 3.9+ can run amory](https://github.com/twosixlabs/armory/issues/1342)
- [armory output record shows version 0.0.0\*](https://github.com/twosixlabs/armory/issues/1569) \[1\]
- [Upgrade ART to 1.12.0](https://github.com/twosixlabs/armory/issues/1581)
- [fix 0.15.4 release build](https://github.com/twosixlabs/armory/issues/1593)


\[1\]. This is caused by the version of armory being run on the host not matching the version being ran in the container. If a new image is not built before `armory run` then `docker/images.py` will pull the most recent image causing a version mismatch. Because older images are unable to pull a version, they default to `0.0.0`; see [`armory/scenarios/main.py`](https://github.com/twosixlabs/armory/blob/master/armory/scenarios/main.py#L211).

File: armory/utils/version.py
Comment (davidslater) [inline armory/utils/version.py:75]: `version_str` as a function argument is never used. Remove.

Comment (davidslater) [reply armory/utils/version.py:75]: Best way to do that is probably using functools
```
import functools
...
@functools.lru_cache
def get_version(...
```

File: armory/utils/version.py
Comment (davidslater) [inline armory/utils/version.py:57]: With multiple reasonable ways of getting the version, I don't think that we want to drop an Error unless we never find a version.

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:55]: `black` and `black[jupyter]` should be bound to version `22.*`. Major version changes will change how it formats.

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:80]: pydub is also needed for `engine` (ART requirement)

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:69]: python-levenshtein should only be needed for deepspeech

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:77]: `datasets` can be grouped into two areas:
1. dataset building. Building certain datasets from scratch requires apache-beam and pydub, but not for typical use. I would not include this part in `pytorch`, etc.
2. dataset loading with TFDS. For this, we need tensorflow-datasets and tensorflow

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:110]: let's put tensorboardx in `engine`

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:102]: `pytorch` should only include `torch`, `torchaudio`, and `torchvision`. The rest of these should be pushed to `deepspeech`

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:108]: what is pycocotools for?

File: pyproject.toml
Comment (davidslater) [reply pyproject.toml:77]: I recommend keeping `datasets` for 2) and creating a `databuilder` or something for 1).

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml:107]: Move protobuf to `datasets`.

**Review by davidslater — CHANGES_REQUESTED** [review]
I'm not done reviewing, but I looked through versioning and dependencies. Will review the rest tomorrow. (I think primarily I have the docker building left.)

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml:108]: It is a image segmentation COCO package. I can find no reference to it in armory or ART. 

For @christopherwoodall, the complex *requirements.txt files have gathered a lot of cruft over the year.

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml:64]: Unreferenced and gone.

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml:69]: done

File: armory/utils/version.py
Comment (christopherwoodall) [reply armory/utils/version.py]: I can update that to a warning.

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml]: unreferenced dependencies removed.

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml]: nope, all have been unpinned

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml]: v1.12.1 was released a couple of days ago. Per your slack, that is the version now.

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml]: per your later comment, these have been put in `datasets-builder` with getting_started.md noting the new target

File: pyproject.toml
Comment (mwartell) [reply pyproject.toml]: moved.

File: armory/utils/version.py
Comment (davidslater) [inline armory/utils/version.py:24]: This line should be:
```
except ImportError:
```
The `importlib` package exists, so it won't raise a `ModuleNotFoundError`

File: armory/utils/version.py
Comment (christopherwoodall) [reply armory/utils/version.py]: Using `setuptools_scm` eliminates the tech debt caused by rolling out our own versioning scheme.

Comment (christopherwoodall) [reply armory/utils/version.py]: I tried using raw options for the regex, but without [modifying `setuptools_scm` directly](https://github.com/pypa/setuptools_scm/blob/main/src/setuptools_scm/version.py#L48) I do not see a way to eliminate the `DISTANCE`, and only return `TAG` and `HASH`.

File: armory/utils/version.py
Comment (mwartell) [reply armory/utils/version.py]: good find, thanks.

File: pyproject.toml
Comment (davidslater) [inline pyproject.toml]: If we're using setuptools for versioning, this should be moved to `dependencies`

Comment (christopherwoodall) [reply pyproject.toml]: This is only required by an older package during the install.

**Review by davidslater — CHANGES_REQUESTED** [review]
Sorry, these were pending comments.

File: .github/workflows/1-scan-lint-build.yml
Comment (davidslater) [inline .github/workflows/1-scan-lint-build.yml:98]: Add `black` checking back in

File: tools/pre-commit.sh
Comment (davidslater) [reply tools/pre-commit.sh:28]: Still no newline - please revert

File: tests/conftest.py
Comment (davidslater) [reply tests/conftest.py:192]: Still no newline - please revert

File: armory/utils/version.py
Comment (davidslater) [inline armory/utils/version.py:47]: Remove the `Path.cwd()` - if you are getting the armory version but are in a non-armory git repo, this may attempt to get the armory version from that repo, unless I understand how the `scm_config` works.

Also, I think `Path(__file__).parent.parent` points to `project_root/armory`, not `project_root`.

File: .github/workflows/1-scan-lint-build.yml
Comment (christopherwoodall) [reply .github/workflows/1-scan-lint-build.yml:61]: No issue. Added back, but this commit will fail the pipeline.

Comment (davidslater) [reply .github/workflows/1-scan-lint-build.yml:61]: Understood. We'll fix after merge.

File: tools/pre-commit.sh
Comment (christopherwoodall) [reply tools/pre-commit.sh:28]: I can [add](https://github.com/twosixlabs/armory/blob/master/tools/pre-commit.sh#L28) a newline, but this will be an edit and not a rollback.

File: tests/conftest.py
Comment (christopherwoodall) [reply tests/conftest.py:192]: This will also be an [edit](https://github.com/twosixlabs/armory/blob/develop/tests/conftest.py).

**Review by davidslater — APPROVED** [review]
LGTM

**Comment by davidslater** [comment]
Will fix linting after merge to not complicate PR.

**Pull request merged**
